torchaudio/functional/functional.py

```diff
@@ -177,12 +177,9 @@ def griffinlim(
     # randomly initialize the phase
     batch, freq, frames = specgram.size()
     if rand_init:
-        angles = 2 * math.pi * torch.rand(batch, freq, frames)
+        angles = torch.rand(batch, freq, frames, dtype=torch.cfloat, device=specgram.device)
```

Comment:
TODO: check to see if `cfloat` is preferred.

Reply:
yeah or even take a step back and consider why we were creating float tensors by default and if it makes sense to  add a dtype `kwarg` that defaults to float

Reply:
@anjali411

I looked into this and the followings are my finding;

1. The use of `cfloat` here is fine when the input Tensor is `float32` and `float64`.
This is because the dtype of `angles` are updated to the correct one in the next `for` loop. Specifically, `specgram * angles` there produces the same dtype as `specgram` so after the first iteration, all the Tensor has the matching dtype as the input.
2. The above suggests that the `griffinlim` will not work for `float16` input type. But I did test and `pow` is not implemented for `float16`. (see the detail bellow) so I do not think it's and issue.

With this finding, I suggest to proceed with the current approach (using `cfloat` for intermediate variables) since it is working for `float32` and `float64`. Also note that `float16` is not in general tested in `torchaudio`.

---

The following is the result of running `griffinlim` for the different dtype on before/after this PR.

### Script
<Details><summary>script</summary>

```python
import torch
import torchaudio.transforms as T


def test(waveform, dtype):
    n_fft = 1024
    win_length = None
    hop_length = 512

    spectrogram = T.Spectrogram(
        n_fft=n_fft,
        win_length=win_length,
        hop_length=hop_length,
    )
    griffin_lim = T.GriffinLim(
        n_fft=n_fft,
        win_length=win_length,
        hop_length=hop_length,
    )

    spec = spectrogram(waveform).to(dtype)
    print('*******')
    print('input:', spec.dtype)
    recon = griffin_lim(spec)
    print('output:', recon.dtype)
    assert recon.dtype == spec.dtype
    return recon


torch.random.manual_seed(0)
waveform = torch.randn(2, 1024)


test(waveform, torch.float64)
test(waveform, torch.float32)
test(waveform, torch.float16)
```

<Details>

### Result
#### Before applying this PR (commit 0433b7aa)

```
*******
input: torch.float64
/home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/functional.py:655: UserWarning: istft will require a complex-valued input tensor in a future PyTorch release. Matching the output from stft with return_complex=True.  (Triggered internally at  /opt/conda/conda-bld/pytorch_1617001512472/work/aten/src/ATen/native/SpectralOps.cpp:807.)
  return _VF.istft(input, n_fft, hop_length, win_length, window, center,  # type: ignore
output: torch.float64
*******
input: torch.float32
output: torch.float32
*******
input: torch.float16
Traceback (most recent call last):
  File "foo.py", line 36, in <module>
    test(waveform, torch.float16)
  File "foo.py", line 24, in test
    recon = griffin_lim(spec)
  File "/home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1015, in _call_impl
    return forward_call(*input, **kwargs)
  File "/scratch/moto/torchaudio/torchaudio/transforms.py", line 189, in forward
    return F.griffinlim(specgram, self.window, self.n_fft, self.hop_length, self.win_length, self.power,
  File "/scratch/moto/torchaudio/torchaudio/functional/functional.py", line 168, in griffinlim
    specgram = specgram.pow(1 / power)
RuntimeError: "pow" not implemented for 'Half'
```

#### After applying this PR

```
*******
input: torch.float64
output: torch.float64
*******
input: torch.float32
output: torch.float32
*******
input: torch.float16
Traceback (most recent call last):
  File "foo.py", line 36, in <module>
    test(waveform, torch.float16)
  File "foo.py", line 24, in test
    recon = griffin_lim(spec)
  File "/home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1015, in _call_impl
    return forward_call(*input, **kwargs)
  File "/scratch/moto/torchaudio/torchaudio/transforms.py", line 189, in forward
    return F.griffinlim(specgram, self.window, self.n_fft, self.hop_length, self.win_length, self.power,
  File "/scratch/moto/torchaudio/torchaudio/functional/functional.py", line 168, in griffinlim
    specgram = specgram.pow(1 / power)
RuntimeError: "pow" not implemented for 'Half'
```

Reply:
This PR https://github.com/pytorch/pytorch/pull/50999 recently added `pow` for `torch.float16`

Reply:
> The use of cfloat here is fine when the input Tensor is float32 and float64.
This is because the dtype of angles are updated to the correct one in the next for loop. Specifically, specgram * angles there produces the same dtype as specgram so after the first iteration, all the Tensor has the matching dtype as the input.

So in that case, I think it makes sense to instead create angles of the same dtype as `specgram` since multiplication in the for loop (`specgram` * `angles`) [promotes all input tensors to the same dtype](https://github.com/pytorch/pytorch/blob/a102adb55e00ea3bc567f6a700d19895f99e148a/aten/src/ATen/TensorIterator.h#L525-L531), so I think it makes sense to create `angles` with dtype `specgram.dtype`.

Reply:
oh this could be the reason why autograd test in #1421 is failing.

Reply:
> This PR [pytorch/pytorch#50999](https://github.com/pytorch/pytorch/pull/50999) recently added `pow` for `torch.float16`

Looks like it's still landing? https://github.com/pytorch/pytorch/pull/55280



> So in that case, I think it makes sense to instead create angles of the same dtype as `specgram` since multiplication in the for loop (`specgram` * `angles`) [promotes all input tensors to the same dtype](https://github.com/pytorch/pytorch/blob/a102adb55e00ea3bc567f6a700d19895f99e148a/aten/src/ATen/TensorIterator.h#L525-L531), so I think it makes sense to create `angles` with dtype `specgram.dtype`.

Added the necessary change.

Reply:
Sorry, #50999 would hopefully reland soon via #55280, and it'd enable `pow` for `float16` on CPU.

`pow` for `float16` is already supported on CUDA.

Reply:
@imaginary-person Thanks for the info.